src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+            } else {
+                first = newNode;
+            }
+            nodeByIndex.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexCheck(index);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexCheck(index);
+        T oldValue = getNode(index).value;
+        getNode(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexCheck(index);
+        Node<T> removedNode = getNode(index);
+        unlink(removedNode, index);
+        size--;
+        return removedNode.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> currentNode = first;
+        int index = 0;
+        for (int i = 0; i < size; i++) {
+            if (object == currentNode.value || object != null && object.equals(currentNode.value)) {
+                break;
+            }
+            currentNode = currentNode.next;
+            index++;
+        }
+        if (currentNode != null) {
+            unlink(currentNode, index);
+            size--;
+            return true;
+        } else {
+            return false;
+        }
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void indexCheck(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(
+                    "Index of value has to be more or equal 0 and less then MyLinkedList size");
+        }
+    }
+
+    public Node<T> getNode(int index) {
+        Node<T> currentNode = null;
+        if (index <= size / 2) {
+            currentNode = first;
+            for (int i = 0; i <= index; i++) {
+                if (i == index) {
+                    break;
+                }
+                currentNode = currentNode.next;
+            }
+        }
+        if (index > size / 2) {
+            currentNode = last;
+            for (int i = size - 1; i >= index; i--) {
+                if (i == index) {
+                    break;
+                }
+                currentNode = currentNode.prev;
+            }
+        }
+        return currentNode;
+    }
+
+    public void unlink(Node<T> removedNode, int index) {
+        if (index > 0) {
+            removedNode.prev.next = removedNode.next;
+        } else {
+            first = removedNode.next;
+        }
+        if (index < size - 1) {
+            removedNode.next.prev = removedNode.prev;
+        } else {
+            last = removedNode.prev;
+        }
     }
```

Comment:
```suggestion
   private T unlink(Node<T> node) {
        Node<T> next = node.next;
        Node<T> prev = node.prev;
        if (next == null) {
            tail = prev;
        } else {
            next.prev = prev;
            node.next = null;
        }
        if (prev == null) {
            head = next;
        } else {
            prev.next = next;
            node.prev = null;
        }
        size--;
    }
```
you dont need index, you can compare references